fastestimator/trace/adapt/terminate_on_nan.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2019 The FastEstimator Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+from typing import Set, Union
+
+import numpy as np
+import tensorflow as tf
+import torch
+
+from fastestimator.schedule.schedule import get_current_items
+from fastestimator.trace.trace import Trace
+from fastestimator.util.data import Data
+from fastestimator.util.traceability_util import traceable
+from fastestimator.util.util import to_set
+
+
+@traceable()
+class TerminateOnNaN(Trace):
+    """End Training if a NaN value is detected. By default (inputs=None) it will monitor all loss values at the end
+    of each batch. If one or more inputs are specified, it will only monitor those values. Inputs may be loss keys
+    and/or the keys corresponding to the outputs of other traces (ex. accuracy).
+
+    Args:
+        monitor_names: key(s) to monitor for NaN values. If None, all loss values will be monitored. "*" will monitor
+            all trace output keys and losses.
+        mode: What mode(s) to execute this Trace in. For example, "train", "eval", "test", or "infer". To execute
+            regardless of mode, pass None. To execute in all modes except for a particular one, you can pass an argument
+            like "!infer" or "!train".
+    """
+    def __init__(self, monitor_names: Union[None, str, Set[str]] = None, mode: Union[None, str,
+                                                                                     Set[str]] = None) -> None:
+        self.monitored_keys = monitor_names if monitor_names is None else to_set(monitor_names)
+        super().__init__(inputs=self.monitored_keys, mode=mode)
+        self.all_loss_keys = {}
+        self.monitored_loss_keys = {}
```

Comment:
self.loss_keys, self.trace_keys can be combined into self.monitor_keys which gets built every epoch. self.monitored_keys can be removed since it is encoded inside self.inputs

Reply:
Sure. I will make this change.